### ISSUE TITLE
v5.0.x computes very different version compared to v4

### ISSUE THREAD

**Issue by belcher-rok** [PaulHatch/semantic-version#84]
After updating to v5.0.1, the Action computes a very different version number.

For example, v5.0.1 computes the next version as `2.1.0-development.90`, whereas v4.0.3 computes the next version as `2.2.25-development.4`.

Our workflow looks like this:
```yaml
      - name: Determine next tag (production)
        if: env.EnvironmentName == 'production'
        id: next_tag_prod
        uses: paulhatch/semantic-version@v5.0.1
        with:
          tag_prefix: ""
          version_format: "${major}.${minor}.${patch}"

      - name: Determine next tag (non-prod)
        if: env.EnvironmentName != 'production'
        id: next_tag_nonprod
        uses: paulhatch/semantic-version@v5.0.1
        with:
          tag_prefix: ""
          version_format: "${major}.${minor}.${patch}-${{ env.EnvironmentName }}.${increment}"
```
----
When I run `git for-each-ref` I see these version tags, many of which were created using v4.0.3:
```txt
$ git for-each-ref --sort=-v:*refname --format="%(refname:short)" --merged=HEAD refs/tags/*[0-9].*[0-9].*[0-9]
2.2.24-development.1
2.2.23-development.4
2.2.22-development.4
2.2.21-development.2
...
2.1.1-development.65
2.1.0-prerelease.1
2.1.0-prerelease.0
2.1.0-development.0
2.0.5
2.0.4
2.0.3
2.0.2
...
```

v5.0.x appears to be using the `2.0.5` tag as the "last release" and is ignoring all the intermediate tags. I'm not sure if this is intentional or not. If so, then perhaps it should also apply **all** the intermediate version bumps, not just the first one. Whatever the case, we are stuck on v4.0.3 for the time being.

Thanks! Let me know if I can provide any extra info or help debug.

**Comment by PaulHatch** [comment]
Yeah, this was fixed recently. Prerelease tags are not supposed to cause the version to increase, they are filtered out now. Based on the output you posted, the last version should be 2.0.5. Since pre-release tags imply that the version was not yet released, so the action will wait until it sees the non-prerelease tag to increase the version.

Previously the tag was obtained using only a call to the git CLI, which uses glob matching and is somewhat limited compared to regex in what can be filtered, resulting in the prerelease tags being selected as the previous tag an prematurely incrementing the version. This list is now filtered in a subsequent step to include only matching tags.

**Comment by alex-rivkin** [comment]
Hi @PaulHatch 
I have an issue related to the original one, but not quite the same.
The logic you've mentioned above works for the example given, but I think it doesn't work as intended if you do have a release between those prerelease tags. In the above example, it's the case where there was a 2.1.1 release before the `2.2.21-development.2` tag.

In my case, I was running the action on a commit that already had a prerelease tag - `v1.1.0-rc9`. I was expecting to get the same tag but I got a strange result - `v1.1.0-rc0`. Looking at the code, I found that the increment number was reset because the current minor was different from the new minor and that has set my increment to zero. 
I found that the mentioned git command - `git for-each-ref --sort=-v:*refname --format="%(refname:short)" --merged=HEAD refs/tags/*[0-9].*[0-9].*[0-9]` produced the following results for me:
```
v1.1.0-rc9
v1.1.0-rc8
v1.1.0-rc4
v1.1.0-rc2
v1.0.0
v1.0.0-rc3
v1.0.0-rc2
v1.0.0-rc1
v0.0.0
```
As you can see, I already have a 1.0.0 release so the next one should be 1.1.0 (I'm setting the `minor_pattern` to `"/.*/"` to get minor default increments for releases).
This led me to find this code:
```
tag = tag
    .split('\n')
    .reverse()
    .find(t => tagFormatter.IsValid(t) && t !== currentTag) || '';
```
Using `reverse()` is what causes the issue - It picks the oldest tag as the last release, and not the latest one. So when deciding on the next version the result is `1.0.0`, which leads to a difference between the last release version and the current version based on the tag of the current commit. This difference resets the increment and returns a different version from the one that is already attached to this commit, while also returning a `false` value for the `changed` output.
I tried some local runs and removing that line indeed fixes the issue, at least in my specific situation. I'm not aware of why the `reverse()` function is needed there, perhaps you can explain. 

Besides the potential bug of calculating a wrong next version, perhaps the explanation to the `changed` output should also include a case in which it will be `false` for running the action on an already tagged commit.

**Comment by PaulHatch** [comment]
Hey @alex-rivkin,
This was due to the fact that the tag validation which excludes prerelease tags was not applied to the tag on the current commit.

This has been fixed in [v5.0.3](https://github.com/PaulHatch/semantic-version/releases/tag/v5.0.3).
